 Udacity-Project: Investigate a Dataset - TMDB-Movies-Data

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction:





In [5]:
# Use this cell to set up import statements for all of the packages that you
#   plan to use.
import pandas as pd
import numpy as np
from datetime import datetime
# Remember to include a 'magic word' so that your visualizations are plotted
#   inline with the notebook. See this page for more:
#   http://ipython.readthedocs.io/en/stable/interactive/magics.html
import matplotlib.pyplot as plt
% matplotlib inline
import seaborn as sns

<a id='wrangling'></a>
## Data Wrangling

> **Tip**: In this section of the report, you will load in the data, check for cleanliness, and then trim and clean your dataset for analysis. Make sure that you document your steps carefully and justify your cleaning decisions.

### General Properties

In [6]:
# Load your data and print out a few lines. Perform operations to inspect data
#   types and look for instances of missing or possibly errant data.
tmdb_data = pd.read_csv('tmdb-movies.csv')

tmdb_data.head()

id    imdb_id  popularity     budget     revenue  \
0  135397  tt0369610   32.985763  150000000  1513528810   
1   76341  tt1392190   28.419936  150000000   378436354   
2  262500  tt2908446   13.112507  110000000   295238201   
3  140607  tt2488496   11.173104  200000000  2068178225   
4  168259  tt2820852    9.335014  190000000  1506249360   

                 original_title  \
0                Jurassic World   
1            Mad Max: Fury Road   
2                     Insurgent   
3  Star Wars: The Force Awakens   
4                     Furious 7   

                                                cast  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...   

                                            homepage          director  \
0                      http://www.jurassicworld.com/   Colin Trevorrow   
1                        http://www.madmaxmovie.com/     George Miller   
2     http://www.thedivergentseries.movie/#insurgent  Robert Schwentke   
3  http://www.starwars.com/films/star-wars-episod...       J.J. Abrams   
4                           http://www.furious7.com/         James Wan   

                         tagline      ...       \
0              The park is open.      ...        
1             What a Lovely Day.      ...        
2     One Choice Can Destroy You      ...        
3  Every generation has a story.      ...        
4            Vengeance Hits Home      ...        

                                            overview runtime  \
0  Twenty-two years after the events of Jurassic ...     124   
1  An apocalyptic story set in the furthest reach...     120   
2  Beatrice Prior must confront her inner demons ...     119   
3  Thirty years after defeating the Galactic Empi...     136   
4  Deckard Shaw seeks revenge against Dominic Tor...     137   

                                      genres  \
0  Action|Adventure|Science Fiction|Thriller   
1  Action|Adventure|Science Fiction|Thriller   
2         Adventure|Science Fiction|Thriller   
3   Action|Adventure|Science Fiction|Fantasy   
4                      Action|Crime|Thriller   

                                production_companies release_date vote_count  \
0  Universal Studios|Amblin Entertainment|Legenda...     6/9/2015       5562   
1  Village Roadshow Pictures|Kennedy Miller Produ...    5/13/2015       6185   
2  Summit Entertainment|Mandeville Films|Red Wago...    3/18/2015       2480   
3          Lucasfilm|Truenorth Productions|Bad Robot   12/15/2015       5292   
4  Universal Pictures|Original Film|Media Rights ...     4/1/2015       2947   

   vote_average  release_year   budget_adj   revenue_adj  
0           6.5          2015  137999939.3  1.392446e+09  
1           7.1          2015  137999939.3  3.481613e+08  
2           6.3          2015  101199955.5  2.716190e+08  
3           7.5          2015  183999919.0  1.902723e+09  
4           7.3          2015  174799923.1  1.385749e+09  

[5 rows x 21 columns]

In [7]:
tmdb_data.shape

(10866, 21)

In [8]:
tmdb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10866 entries, 0 to 10865
Data columns (total 21 columns):
id                      10866 non-null int64
imdb_id                 10856 non-null object
popularity              10866 non-null float64
budget                  10866 non-null int64
revenue                 10866 non-null int64
original_title          10866 non-null object
cast                    10790 non-null object
homepage                2936 non-null object
director                10822 non-null object
tagline                 8042 non-null object
keywords                9373 non-null object
overview                10862 non-null object
runtime                 10866 non-null int64
genres                  10843 non-null object
production_companies    9836 non-null object
release_date            10866 non-null object
vote_count              10866 non-null int64
vote_average            10866 non-null float64
release_year            10866 non-null int64
budget_adj              1

In [9]:
tmdb_data.isnull().sum()

id                         0
imdb_id                   10
popularity                 0
budget                     0
revenue                    0
original_title             0
cast                      76
homepage                7930
director                  44
tagline                 2824
keywords                1493
overview                   4
runtime                    0
genres                    23
production_companies    1030
release_date               0
vote_count                 0
vote_average               0
release_year               0
budget_adj                 0
revenue_adj                0
dtype: int64

> **Tip**: You should _not_ perform too many operations in each cell. Create cells freely to explore your data. One option that you can take with this project is to do a lot of explorations in an initial notebook. These don't have to be organized, but make sure you use enough comments to understand the purpose of each code cell. Then, after you're done with your analysis, create a duplicate notebook where you will trim the excess and organize your steps so that you have a flowing, cohesive report.

> **Tip**: Make sure that you keep your reader informed on the steps that you are taking in your investigation. Follow every code cell, or every set of related code cells, with a markdown cell to describe to the reader what was found in the preceding cell(s). Try to make it so that the reader can then understand what they will be seeing in the following cell(s).

### Data Cleaning: 
Observation:
1. Remove several unused column such as id and imdb_id. Also, other unused columns that has large amount of missing values such as homepage, tagline, keywords, overview, and production companies. 
2. Remove duplicated rows if any.
3. Drop rows with null values.


In [10]:
# After discussing the structure of the data and any problems that need to be
#   cleaned, perform those cleaning steps in the second part of this section.

#deleting unused columns
del_col=['id','imdb_id','homepage','tagline', 'keywords', 'overview', 'production_companies']

#delete the columns
tmdb_data = tmdb_data.drop(del_col,1)

In [11]:
#view the new dataset
tmdb_data.head()

popularity     budget     revenue                original_title  \
0   32.985763  150000000  1513528810                Jurassic World   
1   28.419936  150000000   378436354            Mad Max: Fury Road   
2   13.112507  110000000   295238201                     Insurgent   
3   11.173104  200000000  2068178225  Star Wars: The Force Awakens   
4    9.335014  190000000  1506249360                     Furious 7   

                                                cast          director  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   Colin Trevorrow   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...     George Miller   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...  Robert Schwentke   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...       J.J. Abrams   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...         James Wan   

   runtime                                     genres release_date  \
0      124  Action|Adventure|Science Fiction|Thriller     6/9/2015   
1      120  Action|Adventure|Science Fiction|Thriller    5/13/2015   
2      119         Adventure|Science Fiction|Thriller    3/18/2015   
3      136   Action|Adventure|Science Fiction|Fantasy   12/15/2015   
4      137                      Action|Crime|Thriller     4/1/2015   

   vote_count  vote_average  release_year   budget_adj   revenue_adj  
0        5562           6.5          2015  137999939.3  1.392446e+09  
1        6185           7.1          2015  137999939.3  3.481613e+08  
2        2480           6.3          2015  101199955.5  2.716190e+08  
3        5292           7.5          2015  183999919.0  1.902723e+09  
4        2947           7.3          2015  174799923.1  1.385749e+09

In [12]:
#remove duplicates
tmdb_data.drop_duplicates(inplace=True)

In [13]:
#confirm correction with drop
sum(tmdb_data.duplicated())

0

In [14]:
tmdb_data.shape

(10865, 14)

In [15]:
#drop rows with any null values
tmdb_data.dropna(inplace=True)

In [16]:
#checks if any columns have null values
tmdb_data.isnull().sum().any()

False

In [17]:
tmdb_data.shape

(10731, 14)

<a id='eda'></a>
## Exploratory Data Analysis

> **Tip**: Now that you've trimmed and cleaned your data, you're ready to move on to exploration. Compute statistics and create visualizations with the goal of addressing the research questions that you posed in the Introduction section. It is recommended that you be systematic with your approach. Look at one variable at a time, and then follow it up by looking at relationships between variables.

### Research Question 1: What is the profit for each movie?

In [18]:
# Use this, and more code cells, to explore your data. Don't forget to add
#   Markdown cells to document your observations and findings.

#insert new column for profit value
tmdb_data.insert(3,'profit',tmdb_data['revenue']-tmdb_data['budget'])
tmdb_data.head(1)               


popularity     budget     revenue      profit  original_title  \
0   32.985763  150000000  1513528810  1363528810  Jurassic World   

                                                cast         director  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...  Colin Trevorrow   

   runtime                                     genres release_date  \
0      124  Action|Adventure|Science Fiction|Thriller     6/9/2015   

   vote_count  vote_average  release_year   budget_adj   revenue_adj  
0        5562           6.5          2015  137999939.3  1.392446e+09

Research Question 2: What is the most and least profit.
The most profit is 

In [19]:
#check mean profit of movies
tmdb_data.groupby('profit').mean()
tmdb_data.describe()

popularity        budget       revenue        profit       runtime  \
count  10731.000000  1.073100e+04  1.073100e+04  1.073100e+04  10731.000000   
mean       0.652615  1.480365e+07  4.031989e+07  2.551624e+07    102.468829   
std        1.004804  3.106456e+07  1.176524e+08  9.715071e+07     30.493873   
min        0.000188  0.000000e+00  0.000000e+00 -4.139124e+08      0.000000   
25%        0.210765  0.000000e+00  0.000000e+00  0.000000e+00     90.000000   
50%        0.387081  0.000000e+00  0.000000e+00  0.000000e+00     99.000000   
75%        0.720889  1.600000e+07  2.500000e+07  9.971094e+06    112.000000   
max       32.985763  4.250000e+08  2.781506e+09  2.544506e+09    900.000000   

         vote_count  vote_average  release_year    budget_adj   revenue_adj  
count  10731.000000  10731.000000  10731.000000  1.073100e+04  1.073100e+04  
mean     219.812972      5.964710   2001.259622  1.776530e+07  5.200623e+07  
std      578.815324      0.930283     12.820151  3.446630e+07  1.454252e+08  
min       10.000000      1.500000   1960.000000  0.000000e+00  0.000000e+00  
25%       17.000000      5.400000   1995.000000  0.000000e+00  0.000000e+00  
50%       39.000000      6.000000   2006.000000  0.000000e+00  0.000000e+00  
75%      148.000000      6.600000   2011.000000  2.110885e+07  3.470546e+07  
max     9767.000000      9.200000   2015.000000  4.250000e+08  2.827124e+09

Research Question 3: Which movies with most and least profit.

In [20]:
#define the function
def calculate(column):
    #high profit
    high=tmdb_data[column].idxmax()
    high_data=pd.DataFrame(tmdb_data.loc[high])
    
    #low profit
    low=tmdb_data[column].idxmin()
    low_data=pd.DataFrame(tmdb_data.loc[low])
    
    #collect data
    data=pd.concat([high_data, low_data], axis=1)
    
    return data

calculate ('profit')
    

1386  \
popularity                                                9.43277   
budget                                                  237000000   
revenue                                                2781505847   
profit                                                 2544505847   
original_title                                             Avatar   
cast            Sam Worthington|Zoe Saldana|Sigourney Weaver|S...   
director                                            James Cameron   
runtime                                                       162   
genres                   Action|Adventure|Fantasy|Science Fiction   
release_date                                           12/10/2009   
vote_count                                                   8458   
vote_average                                                  7.1   
release_year                                                 2009   
budget_adj                                            2.40887e+08   
revenue_adj                                           2.82712e+09   

                                                             2244  
popularity                                                0.25054  
budget                                                  425000000  
revenue                                                  11087569  
profit                                                 -413912431  
original_title                                  The Warrior's Way  
cast            Kate Bosworth|Jang Dong-gun|Geoffrey Rush|Dann...  
director                                               Sngmoo Lee  
runtime                                                       100  
genres                  Adventure|Fantasy|Action|Western|Thriller  
release_date                                            12/2/2010  
vote_count                                                     74  
vote_average                                                  6.4  
release_year                                                 2010  
budget_adj                                               4.25e+08  
revenue_adj                                           1.10876e+07

As you can see the original title movie 'Avatar' has the most profit of 2544505847, and the movie 'The Warrior's Way' has the least profit of -413912431.

### Research Question 4: Genres Category

In [21]:
# Continue to explore the data to address your additional research
#   questions. Add more headers as needed if you have more questions to
#   investigate.

#genres have more than one category and have string that holds multiple values. We need to separate the string.
tmdb_data.drop('genres', axis=1).join(tmdb_data['genres'].str.split('|', expand=True).stack().reset_index(level=1, drop=True).rename('genres'))


popularity     budget     revenue      profit  \
0       32.985763  150000000  1513528810  1363528810   
0       32.985763  150000000  1513528810  1363528810   
0       32.985763  150000000  1513528810  1363528810   
0       32.985763  150000000  1513528810  1363528810   
1       28.419936  150000000   378436354   228436354   
1       28.419936  150000000   378436354   228436354   
1       28.419936  150000000   378436354   228436354   
1       28.419936  150000000   378436354   228436354   
2       13.112507  110000000   295238201   185238201   
2       13.112507  110000000   295238201   185238201   
2       13.112507  110000000   295238201   185238201   
3       11.173104  200000000  2068178225  1868178225   
3       11.173104  200000000  2068178225  1868178225   
3       11.173104  200000000  2068178225  1868178225   
3       11.173104  200000000  2068178225  1868178225   
4        9.335014  190000000  1506249360  1316249360   
4        9.335014  190000000  1506249360  1316249360   
4        9.335014  190000000  1506249360  1316249360   
5        9.110700  135000000   532950503   397950503   
5        9.110700  135000000   532950503   397950503   
5        9.110700  135000000   532950503   397950503   
5        9.110700  135000000   532950503   397950503   
6        8.654359  155000000   440603537   285603537   
6        8.654359  155000000   440603537   285603537   
6        8.654359  155000000   440603537   285603537   
6        8.654359  155000000   440603537   285603537   
7        7.667400  108000000   595380321   487380321   
7        7.667400  108000000   595380321   487380321   
7        7.667400  108000000   595380321   487380321   
8        7.404165   74000000  1156730962  1082730962   
...           ...        ...         ...         ...   
10853    0.163592          0           0           0   
10853    0.163592          0           0           0   
10854    0.146402          0           0           0   
10854    0.146402          0           0           0   
10854    0.146402          0           0           0   
10855    0.141026     700000           0     -700000   
10855    0.141026     700000           0     -700000   
10855    0.141026     700000           0     -700000   
10855    0.141026     700000           0     -700000   
10856    0.140934          0           0           0   
10856    0.140934          0           0           0   
10856    0.140934          0           0           0   
10857    0.131378          0           0           0   
10857    0.131378          0           0           0   
10858    0.317824          0           0           0   
10858    0.317824          0           0           0   
10859    0.089072          0           0           0   
10859    0.089072          0           0           0   
10859    0.089072          0           0           0   
10859    0.089072          0           0           0   
10860    0.087034          0           0           0   
10861    0.080598          0           0           0   
10862    0.065543          0           0           0   
10862    0.065543          0           0           0   
10862    0.065543          0           0           0   
10863    0.065141          0           0           0   
10863    0.065141          0           0           0   
10864    0.064317          0           0           0   
10864    0.064317          0           0           0   
10865    0.035919      19000           0      -19000   

                                         original_title  \
0                                        Jurassic World   
0                                        Jurassic World   
0                                        Jurassic World   
0                                        Jurassic World   
1                                    Mad Max: Fury Road   
1                                    Mad Max: Fury Road   
1                                    Mad Max: Fury Road   
1                                    Mad Max: Fury Road   
2  

In [22]:
tmdb_data['genres'].value_counts()


Drama                                                 710
Comedy                                                707
Drama|Romance                                         289
Comedy|Drama                                          280
Documentary                                           274
Comedy|Romance                                        267
Horror|Thriller                                       259
Horror                                                253
Comedy|Drama|Romance                                  222
Drama|Thriller                                        138
Comedy|Family                                         102
Action|Thriller                                       101
Thriller                                               93
Drama|Comedy                                           92
Animation|Family                                       84
Crime|Drama|Thriller                                   81
Crime|Drama                                            74
Comedy|Horror 

<a id='conclusions'></a>
## Conclusions

> **Tip**: Finally, summarize your findings and the results that have been performed. Make sure that you are clear with regards to the limitations of your exploration. If you haven't done any statistical tests, do not imply any statistical conclusions. And make sure you avoid implying causation from correlation!

> **Tip**: Once you are satisfied with your work here, check over your report to make sure that it is satisfies all the areas of the rubric (found on the project submission page at the end of the lesson). You should also probably remove all of the "Tips" like this one so that the presentation is as polished as possible.

## Submitting your Project 

> Before you submit your project, you need to create a .html or .pdf version of this notebook in the workspace here. To do that, run the code cell below. If it worked correctly, you should get a return code of 0, and you should see the generated .html file in the workspace directory (click on the orange Jupyter icon in the upper left).

> Alternatively, you can download this report as .html via the **File** > **Download as** submenu, and then manually upload it into the workspace directory by clicking on the orange Jupyter icon in the upper left, then using the Upload button.

> Once you've done this, you can submit your project by clicking on the "Submit Project" button in the lower right here. This will create and submit a zip file with this .ipynb doc and the .html or .pdf version you created. Congratulations!

In [ ]:
from subprocess import call
call(['python', '-m', 'nbconvert', 'Investigate_a_Dataset.ipynb'])